### DISCUSSION TITLE
Planning for mamaki upgrades

### DISCUSSION THREAD

**Discussion by evan-forbes** [celestiaorg/celestia-app#543] [General]
There are several factors that could affect when and how we upgrade mamaki. We should plan and discuss those things here. Ideally we can reach consensus over which features can go in which upgrades.

off the top of my head
- we will need to hardfork when we upgrade the cosmos-sdk/downgrade to tendermint v0.34
- we will need to hardfork when we introduce non-interactive defaults https://github.com/celestiaorg/celestia-app/issues/382
- we will need to hardfork when we pay for padding by default #431
- we will need to hardfork when we introduce the vesting module https://github.com/celestiaorg/celestia-app/issues/280
- we will need to hardfork when we introduce the QGB
- we will need to hardfork when we pick different parameters https://github.com/celestiaorg/celestia-app/issues/169
- we will need to hardfork when we re-enable IBC/IBC allowlist https://github.com/celestiaorg/celestia-app/issues/235 https://github.com/celestiaorg/celestia-app/issues/412
- we will likely restart the chain when introducing 512 byte shares https://github.com/celestiaorg/celestia-core/issues/724 and using leopard/go-leopard https://github.com/celestiaorg/celestia-core/issues/673

Each of the above upgrades could be combined or kept separate.

Assuming that we can get mamaki to a stable enough point, either with v0.6.0 or with another non-hardfork upgrade, then we have more freedom on when we can upgrade.

**Comment by liamsi** [comment]
> we will need to hardfork when we upgrade the cosmos-sdk/downgrade to tendermint v0.34

I still do feel we should do this as soon as possible and independent of the other changes. I want the testnet to be as stable as possible again and not wait for additional features.

**Reply by evan-forbes** [reply]
That makes sense and I agree assuming that v0.6.0 doesn't give us very consistent block times. If we get very consistent block times and no down time, then I think we have more freedom.

**Reply by evan-forbes** [reply]
if we upgrade ASAP, then we will have to have at least 3 upgrades, assuming that we don't want to bundle everything else into one upgrade.

**Reply by liamsi** [reply]
I do not mind 3 upgrades if 2 of those are done via the upgrade module and gov voting. As the coordination would happen on chain.

**Reply by liamsi** [reply]
@musalbas might disagree here.

**Reply by musalbas** [reply]
I think this should be discussed with #validators at least in Discord. 3 upgrades might require quite a lot of coordination.

**Reply by evan-forbes** [reply]
> if 2 of those are done via the upgrade module and gov voting.

I'd need to try it out to be sure, but since we can run arbitrary logic in the upgrade handler, we might be able to get away with a chain reset using the normal gov + upgrade modules.

**Comment by adlerjohn** [comment]
The non-interactive defaults aren't a consensus rule, and shouldn't require a hard fork. Could you clarify why they would require a hard fork, I might be missing something.

**Reply by evan-forbes** [reply]
sorry that I missed this comment, it will need a hardfork mainly because of wrapped txs. Both because nodes that don't know how to interpret them will ignore otherwise valid txs, and because this changes how message inclusion is determined.

**Comment by adlerjohn** [comment]
The QGB can be introduced as a soft fork, not a hard fork. It requires validator nodes to run addition logic that makes what would have previously been valid blocks, invalid. But it doesn't allow any blocks that would have been invalid to be considered valid.

Edit: nevermind, I forgot that the QGB in its current incarnation involves app state.

**Comment by evan-forbes** [comment]
closing because this is out of date. we've since hardspooned to mocha
